File: src/main/scala/DTAPages.scala
package vtpassim

import org.apache.spark.sql.SparkSession

import scala.collection.mutable.{ListBuffer, Stack, StringBuilder}
import scala.util.Try
import scala.xml.pull._

case class RenditionSpan(rendition: String, start: Int, length: Int)

case class ZoneInfo(ztype: String, place: String)

case class ZoneContent(info: ZoneInfo, data: StringBuilder, rend: ListBuffer[RenditionSpan])

case class Rec(id: String, book: String, seq: Int, page: String, ztype: String, place: String,
  text:String, rendition: Array[RenditionSpan])

object DTAPages {
  def main(args: Array[String]) {
    val spark = SparkSession.builder().appName("DTAPages Import").getOrCreate()
    import spark.implicits._

    spark.sparkContext.hadoopConfiguration
      .set("mapreduce.input.fileinputformat.input.dir.recursive", "true")

    val lines = Seq("head")
    val floats = Seq("figure", "note", "table")

    spark.sparkContext.binaryFiles(args(0), spark.sparkContext.defaultParallelism)
      .filter(_._1.endsWith(".xml"))
      .flatMap( in => {
        val fname = new java.io.File(new java.net.URL(in._1).toURI)
        val book = fname.getName.replaceAll("(.TEI-P5)?.xml$", "")
        var seq = -1
        var pageID = ""
        val zoneStack = new Stack[ZoneContent]()
        val rendStack = new Stack[RenditionSpan]()

        val pass = new XMLEventReader(scala.io.Source.fromURL(in._1))
        pass.flatMap { event =>
          event match {
            case EvElemStart(_, elem, attr, _) if lines.contains(elem) && !zoneStack.isEmpty => {
              rendStack.push(RenditionSpan(elem, zoneStack.top.data.length, 0))
              Nil
            }
            case EvElemEnd(_, elem) if lines.contains(elem) && !zoneStack.isEmpty => {
              val start = rendStack.pop
              zoneStack.top.rend += RenditionSpan(start.rendition, start.start,
                zoneStack.top.data.length - start.start)
              Nil
            }
            case EvElemStart(_, "hi", attr, _) if !zoneStack.isEmpty => {
              rendStack.push(RenditionSpan(Try(attr("rendition").text).getOrElse(""),
                zoneStack.top.data.length, 0))
              Nil
            }
            case EvElemEnd(_, "hi") if !zoneStack.isEmpty => {
              val start = rendStack.pop
              zoneStack.top.rend ++= start.rendition.split("\\s+")
                .filter { _ != "" }
                .map { r => RenditionSpan(r.stripPrefix("#"),
                  start.start, zoneStack.top.data.length - start.start) }
              Nil
            }
            case EvElemStart(_, "pb", attr, _) => {
              // Record and restart all open rendition spans
              for ( i <- 0 until rendStack.length ) {
                zoneStack.top.rend += RenditionSpan(rendStack(i).rendition, rendStack(i).start,
                  zoneStack.top.data.length - rendStack(i).start)
                rendStack(i) = RenditionSpan(rendStack(i).rendition, 0, 0)
              }
              val res = new ListBuffer[Rec]()
              // Remember and output all open zones
              val zones = (if ( zoneStack.isEmpty ) Seq(ZoneInfo("body", "body")) else zoneStack.toSeq.map(_.info).reverse)
              while ( !zoneStack.isEmpty ) {
                seq += 1
                val top = zoneStack.pop
                res += Rec(pageID + s"z$seq", book, seq, pageID, top.info.ztype, top.info.place, top.data.toString, top.rend.toArray)
              }
              for ( zone <- zones ) {
                zoneStack.push(new ZoneContent(zone, new StringBuilder, new ListBuffer[RenditionSpan]()))
              }
              // Output printed page number (n attribute not in brackets) here?
              pageID = book + attr("facs").text
              val pno = Try(attr("n").text).getOrElse("").replaceAll("\\[[^\\]]+\\]", "")
              if ( pno != "" ) {
                seq += 1
                res += Rec(pageID + s"z$seq", book, seq, pageID, "pageNum", "pageNum", pno, new Array[RenditionSpan](0))
              }
              res
            }
            case EvElemStart(_, "cb", attr, _) => {
              // Record and restart all open rendition spans
              for ( i <- 0 until rendStack.length ) {
                zoneStack.top.rend += RenditionSpan(rendStack(i).rendition, rendStack(i).start,
                  zoneStack.top.data.length - rendStack(i).start)
                rendStack(i) = RenditionSpan(rendStack(i).rendition, 0, 0)
              }
              val res = new ListBuffer[Rec]()
              // Remember and output all open zones
              val zones = (if ( zoneStack.isEmpty ) Seq(ZoneInfo("body", "body")) else zoneStack.toSeq.map(_.info).reverse)
              while ( !zoneStack.isEmpty ) {
                val top = zoneStack.pop
                if ( top.data.toString != "" ) {
                  seq += 1
                  res += Rec(pageID + s"z$seq", book, seq, pageID, top.info.ztype, top.info.place, top.data.toString, top.rend.toArray)
                }
              }
              for ( zone <- zones ) {
                zoneStack.push(new ZoneContent(zone, new StringBuilder, new ListBuffer[RenditionSpan]()))
              }
              // Output printed column number (n attribute not in brackets) here?
              val cno = Try(attr("n").text).getOrElse("").replaceAll("\\[[^\\]]+\\]", "")
              if ( cno != "" ) {
                seq += 1
                res += Rec(pageID + s"z$seq", book, seq, pageID, "colNum", "colNum", cno, new Array[RenditionSpan](0))
              }
              res
            }
            case EvElemEnd(_, "text") => {
              val res = new ListBuffer[Rec]()
              while ( !zoneStack.isEmpty ) {
                seq += 1
                val top = zoneStack.pop
                res += Rec(pageID + s"z$seq", book, seq, pageID, top.info.ztype, top.info.place, top.data.toString, top.rend.toArray)
              }
              res
            }
            case EvElemStart(_, elem, attr, _) if floats.contains(elem) && !zoneStack.isEmpty => {
              zoneStack.top.rend += RenditionSpan(elem, zoneStack.top.data.length, 0)
              zoneStack.push(new ZoneContent(ZoneInfo(elem,
                Try(attr("place").text).getOrElse(elem)),
                new StringBuilder, new ListBuffer[RenditionSpan]()))
              Nil
            }
            case EvElemEnd(_, elem) if floats.contains(elem) && !zoneStack.isEmpty => {
              val top = zoneStack.pop
              seq += 1
              Seq(Rec(pageID + s"z$seq", book, seq, pageID, top.info.ztype, top.info.place, top.data.toString, top.rend.toArray))
            }
            case EvElemStart(_, "fw", attr, _) if !zoneStack.isEmpty => {
              zoneStack.push(new ZoneContent(ZoneInfo(Try(attr("type").text).getOrElse("fw"),
                Try(attr("place").text).getOrElse("fw")),
                new StringBuilder, new ListBuffer[RenditionSpan]()))
              Nil
            }
            case EvElemEnd(_, "fw") if !zoneStack.isEmpty => {
              val top = zoneStack.pop
              seq += 1
              Seq(Rec(pageID + s"z$seq", book, seq, pageID, top.info.ztype, top.info.place, top.data.toString, top.rend.toArray))
            }
            case EvElemEnd(_, "cell") if !zoneStack.isEmpty => {
              zoneStack.top.data ++= "\t"
              Nil
            }
            case EvText(t) if !zoneStack.isEmpty => {
              // remove leading whitespace only if we haven't added anything
              zoneStack.top.data ++= (if (zoneStack.top.data.isEmpty) t.replaceAll("^\\s+", "") else t)
              Nil
            }
            case EvEntityRef(n) if !zoneStack.isEmpty => {
              zoneStack.top.data ++= (n match {
                case "amp" => "&"
                case "lt" => "<"
                case "gt" => ">"
                case "apos" => "'"
                case "quot" => "\""
                case _ => ""
              })
              Nil
            }
            case _ => Nil
          }
        }
      })
      .toDF
      .write.save(args(1))
    spark.stop()
  }
}

File: scripts/dta-pages.py
import argparse, os, re
from re import sub
from pyspark.sql import SparkSession, Row
from pyspark.sql.functions import col, collect_list, explode, sort_array, struct, udf
import pyspark.sql.functions as f
from io import StringIO, BytesIO
from lxml import etree
from dataclasses import dataclass

ns = {None: 'http://www.tei-c.org/ns/1.0'}

@dataclass
class RenditionSpan:
    rendition: str
    start: int
    length: int

@dataclass
class ZoneInfo:
    ztype: str
    place: str

@dataclass
class ZoneContent:
    info: ZoneInfo
    data: str
    rend: list

@dataclass
class Rec:
    id: str
    book: str
    seq: int
    page: str
    ztype: str
    place: str
    text: str
    rendition: list

class BookStream(object):
    lines = set(['head'])
    floats = set(['figure', 'note', 'table'])
    def __init__(self, book):
        self.book = book
        self.seq = -1
        self.pageID = ''
        self.zones = []
        self.rends = []
        self.res = []
    def start(self, elem, attrib):
        tag = etree.QName(elem).localname
        if tag == 'hi' and len(self.zones) > 0:
            self.rends.append(RenditionSpan(attrib.get('rendition', ''),
                                            len(self.zones[-1].data), 0))
        elif tag == 'fw' and len(self.zones) > 0:
            self.zones.append(ZoneContent(ZoneInfo(attrib.get('type', 'fw'),
                                                   attrib.get('place', 'fw')), '', []))
        elif tag == 'gap' and len(self.zones) > 0:
            self.zones.append(ZoneContent(ZoneInfo('quash', 'quash'), '', []))
        elif tag == 'g' and len(self.zones) > 0:
            if attrib.get('ref', '') == 'char:EOLhyphen':
                self.zones[-1].data += '\u00ad\n'
        elif tag == 'pb':
            # Record and restart all open rendition spans
            for i in range(len(self.rends)):
                r = self.rends[i]
                self.zones[-1].rend.append(RenditionSpan(r.rendition, r.start,
                                                         len(self.zones[-1].data) - r.start))
                self.rends[i] = RenditionSpan(r.rendition, 0, 0)
            # Remember and output all open zones
            if len(self.zones) > 0:
                zinfo = [r.info for r in reversed(self.zones)]
                while len(self.zones) > 0:
                    self.seq += 1
                    top = self.zones.pop()
                    self.res.append(Row(id=f'{self.pageID}z{self.seq}', book=self.book,
                                        seq=self.seq, page=self.pageID,
                                        ztype=top.info.ztype, place=top.info.place,
                                        text=top.data, rendition=top.rend))
            else:
                zinfo = [ZoneInfo('body', 'body')]
            for z in zinfo:
                self.zones.append(ZoneContent(z, '', []))
                # Output printed page number (n attribute not in brackets) here
                self.pageID = self.book + attrib.get('facs', '')
                pno = re.sub(r'\[[^\]]+\]', '', attrib.get('n', ''))
            if pno != '':
                self.seq += 1
                self.res.append(Row(id=f'{self.pageID}z{self.seq}', book=self.book,
                                    seq=self.seq, page=self.pageID,
                                    ztype='pageNum', place='pageNum', text=pno, rendition=[]))
        elif tag =='cb':
            # Record and restart all open rendition spans
            for i in range(len(self.rends)):
                r = self.rends[i]
                self.zones[-1].rend.append(RenditionSpan(r.rendition, r.start,
                                                         len(self.zones[-1].data) - r.start))
                self.rends[i] = RenditionSpan(r.rendition, 0, 0)
            # Remember and output all open zones
            if len(self.zones) > 0:
                zinfo = [r.info for r in reversed(self.zones)]
                while len(self.zones) > 0:
                    top = self.zones.pop()
                    if len(top.data) > 0:
                        self.seq += 1
                        self.res.append(Row(id=f'{self.pageID}z{self.seq}', book=self.book,
                                            seq=self.seq, page=self.pageID,
                                            ztype=top.info.ztype, place=top.info.place,
                                            text=top.data, rendition=top.rend))
            else:
                zinfo = [ZoneInfo('body', 'body')]
            for z in zinfo:
                self.zones.append(ZoneContent(z, '', []))
            # Output printed column number (n attribute not in brackets) here
            cno = re.sub(r'\[[^\]]+\]', '', attrib.get('n', ''))
            if cno != '':
                self.seq += 1
                self.res.append(Row(id=f'{self.pageID}z{self.seq}', book=self.book,
                                    seq=self.seq, page=self.pageID,
                                    ztype='colNum', place='colNum', text=cno, rendition=[]))
        elif tag in self.lines and len(self.zones) > 0:
            self.rends.append(RenditionSpan(tag, len(self.zones[-1].data), 0))
        elif tag in self.floats and len(self.zones) > 0:
            self.zones[-1].rend.append(RenditionSpan(tag, len(self.zones[-1].data), 0))
            self.zones.append(ZoneContent(ZoneInfo(tag, attrib.get('place', tag)), '', []))
            
    def end(self, elem):
        tag = etree.QName(elem).localname
        if tag == 'hi' and len(self.zones) > 0:
            start = self.rends.pop()
            self.zones[-1].rend += [RenditionSpan(r.lstrip('#'), start.start,
                                                  len(self.zones[-1].data) - start.start)
                                    for r in re.split(r'\s+', start.rendition) if r != '']
        elif tag == 'fw' and len(self.zones) > 0:
            top = self.zones.pop()
            self.seq += 1
            self.res.append(Row(id=f'{self.pageID}z{self.seq}', book=self.book,
                                seq=self.seq, page=self.pageID,
                                ztype=top.info.ztype, place=top.info.place,
                                text=top.data, rendition=top.rend))
        elif tag == 'gap' and len(self.zones) > 0:
            top = self.zones.pop()
            self.zones[-1].data += sub(r'^.non-Latin.*$', '@@@',
                                       sub(r'\s+$', '', sub(r'^\s+', '', top.data)))
        elif tag == 'cell' and len(self.zones) > 0:
            self.zones[-1].data += '\t'
        elif tag == 'text':
            while len(self.zones) > 0:
                self.seq += 1
                top = self.zones.pop()
                self.res.append(Row(id=f'{self.pageID}z{self.seq}', book=self.book,
                                    seq=self.seq, page=self.pageID,
                                    ztype=top.info.ztype, place=top.info.place,
                                    text=top.data, rendition=top.rend))
        elif tag in self.lines and len(self.zones) > 0:
            start = self.rends.pop()
            self.zones[-1].rend.append(RenditionSpan(start.rendition, start.start,
                                                     len(self.zones[-1].data) - start.start))
        elif tag in self.floats and len(self.zones) > 0:
            top = self.zones.pop()
            self.seq += 1
            self.res.append(Row(id=f'{self.pageID}z{self.seq}', book=self.book,
                                seq=self.seq, page=self.pageID,
                                ztype=top.info.ztype, place=top.info.place,
                                text=top.data, rendition=top.rend))

    def data(self, data):
        if len(self.zones) > 0:
            # remove leading whitespace only if we haven't added anything
            if len(self.zones[-1].data) == 0:
                data = re.sub(r'^\s+', '', data)
            self.zones[-1].data += re.sub(r'\n[ ]+', '\n', data)
    def comment(self, text):
        pass
    def close(self):
        res = self.res
        self.seq = -1
        self.pageID = ''
        self.zones = []
        self.rends = []
        self.res = []
        return res

def parseFile(path, content):
    book = re.sub(r'(.TEI-P5)?.xml$', '', os.path.basename(path))

    parser = etree.XMLParser(target = BookStream(book))
    result = etree.parse(BytesIO(content), parser)
    return result

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='DTA Pages',
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('inputPath', metavar='<input path>', help='input path')
    parser.add_argument('outputPath', metavar='<output path>', help='output path')

    config = parser.parse_args()

    spark = SparkSession.builder.appName('DTA Pages').getOrCreate()

    spark.read.load(config.inputPath, format='binaryFile', recursiveFileLookup='true',
                    pathGlobFilter='*.xml'
        ).rdd.flatMap(lambda r: parseFile(r.path, r.content)
        ).toDF(
        ).write.json(config.outputPath, mode='overwrite')

    spark.stop()

                    
